File: python/programmers/level 2/스킬트리.py
def solution(skill, skill_trees):
    answer = 0
    
    for skill_tree in skill_trees:
        skill_copy = skill
        for s in skill_tree:
            if s in skill_copy:
                if s != skill_copy[0]:
                    answer -= 1
                    break
                else:
                    skill_copy = skill_copy[1:]
        answer += 1
    return answer

File: java/src/main/java/ps_traning/programmers/level_2/스킬트리.java
package ps_traning.programmers.level_2;

import java.util.HashSet;
import java.util.Set;

public class 스킬트리 {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        Set<Character> set = new HashSet<>();
        for (char s : skill.toCharArray()) set.add(s);
        for (String st : skill_trees) {
            StringBuilder sb = new StringBuilder(skill);
            sb = sb.reverse();
            int index = sb.length() - 1;
            for (char s : st.toCharArray()) {
                if (set.contains(s)) {
                    if (sb.charAt(index) == s) sb.setLength(index--);
                    else {
                        answer--;
                        break;
                    }
                }
            }
            answer++;
        }
        return answer;
    }
}
